File: Net/GraphicsPipeline/DynamicBiking/VBNet/BikingTrackModeCmd.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Drawing
Imports System.Runtime.InteropServices
Imports ESRI.ArcGIS.ADF.BaseClasses
Imports ESRI.ArcGIS.ADF.CATIDs
Imports ESRI.ArcGIS.Controls
Imports ESRI.ArcGIS.Carto
Imports ESRI.ArcGIS.Display
Imports ESRI.ArcGIS.esriSystem
Imports ESRI.ArcGIS.SystemUI

	''' <summary>
	''' Summary description for BikingTrackModeCmd.
	''' </summary>
	<Guid("5a26e262-9e4c-498f-b77c-a6fdeee0dd4b"), ClassInterface(ClassInterfaceType.None), ProgId("BikingTrackModeCmd")> _
	Public NotInheritable Class BikingTrackModeCmd : Inherits BaseCommand
		#Region "COM Registration Function(s)"
		<ComRegisterFunction(), ComVisible(False)> _
		Private Shared Sub RegisterFunction(ByVal registerType As Type)
			' Required for ArcGIS Component Category Registrar support
			ArcGISCategoryRegistration(registerType)

			'
			' TODO: Add any COM registration code here
			'
		End Sub

		<ComUnregisterFunction(), ComVisible(False)> _
		Private Shared Sub UnregisterFunction(ByVal registerType As Type)
			' Required for ArcGIS Component Category Registrar support
			ArcGISCategoryUnregistration(registerType)

			'
			' TODO: Add any COM unregistration code here
			'
		End Sub

		#Region "ArcGIS Component Category Registrar generated code"
		''' <summary>
		''' Required method for ArcGIS Component Category registration -
		''' Do not modify the contents of this method with the code editor.
		''' </summary>
		Private Shared Sub ArcGISCategoryRegistration(ByVal registerType As Type)
			Dim regKey As String = String.Format("HKEY_CLASSES_ROOT\CLSID\{{{0}}}", registerType.GUID)
			ControlsCommands.Register(regKey)

		End Sub
		''' <summary>
		''' Required method for ArcGIS Component Category unregistration -
		''' Do not modify the contents of this method with the code editor.
		''' </summary>
		Private Shared Sub ArcGISCategoryUnregistration(ByVal registerType As Type)
			Dim regKey As String = String.Format("HKEY_CLASSES_ROOT\CLSID\{{{0}}}", registerType.GUID)
			ControlsCommands.Unregister(regKey)

		End Sub

		#End Region
		#End Region

		Private m_hookHelper As IHookHelper = Nothing

		Private m_dynamicBikingCmd As DynamicBikingCmd = Nothing

		Public Sub New()
			MyBase.m_category = ".NET Samples"
			MyBase.m_caption = "Dynamic biking track"
			MyBase.m_message = "Dynamic biking track mode"
			MyBase.m_toolTip = "Dynamic biking track mode"
			MyBase.m_name = "BikingTrackModeCmd"

			Try
				Dim bitmapResourceName As String = Me.GetType().Name & ".bmp"
				MyBase.m_bitmap = New Bitmap(Me.GetType(), bitmapResourceName)
			Catch ex As Exception
				System.Diagnostics.Trace.WriteLine(ex.Message, "Invalid Bitmap")
			End Try
		End Sub

#Region "Overridden Class Methods"

    ''' <summary>
    ''' Occurs when this command is created
    ''' </summary>
    ''' <param name="hook">Instance of the application</param>
    Public Overrides Sub OnCreate(ByVal hook As Object)
        If hook Is Nothing Then
            Return
        End If

        If m_hookHelper Is Nothing Then
            m_hookHelper = New HookHelperClass()
        End If

        m_hookHelper.Hook = hook
    End Sub

    ''' <summary>
    ''' Occurs when this command is clicked
    ''' </summary>
    Public Overrides Sub OnClick()
        If Not m_dynamicBikingCmd Is Nothing Then
            m_dynamicBikingCmd.TrackMode = Not m_dynamicBikingCmd.TrackMode
        End If
    End Sub

    Public Overrides ReadOnly Property Enabled() As Boolean
        Get
            m_dynamicBikingCmd = GetBikingCmd()
            If Not m_dynamicBikingCmd Is Nothing Then
                Return m_dynamicBikingCmd.IsPlaying
            End If

            Return False
        End Get
    End Property

    Public Overrides ReadOnly Property Checked() As Boolean
        Get
            If Not m_dynamicBikingCmd Is Nothing Then
                Return m_dynamicBikingCmd.TrackMode
            End If
            Return False
        End Get
    End Property

#End Region

		Private Function GetBikingCmd() As DynamicBikingCmd
			If m_hookHelper.Hook Is Nothing Then
				Return Nothing
			End If

			Dim dynamicBikingCmd As DynamicBikingCmd = Nothing
			If TypeOf m_hookHelper.Hook Is IToolbarControl2 Then
				Dim toolbarCtrl As IToolbarControl2 = CType(m_hookHelper.Hook, IToolbarControl2)
				Dim commandPool As ICommandPool2 = TryCast(toolbarCtrl.CommandPool, ICommandPool2)
				Dim commantCount As Integer = commandPool.Count
				Dim command As ICommand = Nothing
				Dim i As Integer = 0
				Do While i < commantCount
					command = commandPool.Command(i)
					If TypeOf command Is DynamicBikingCmd Then
						dynamicBikingCmd = CType(command, DynamicBikingCmd)
					End If
					i += 1
				Loop
			End If

			Return dynamicBikingCmd
		End Function
	End Class

File: Net/GraphicsPipeline/DynamicBiking/CSharp/BikingTrackModeCmd.cs
/*

   Copyright 2019 Esri

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.

   See the License for the specific language governing permissions and
   limitations under the License.

*/
using System;
using System.Drawing;
using System.Runtime.InteropServices;
using ESRI.ArcGIS.ADF.BaseClasses;
using ESRI.ArcGIS.ADF.CATIDs;
using ESRI.ArcGIS.Controls;
using ESRI.ArcGIS.Carto;
using ESRI.ArcGIS.Display;
using ESRI.ArcGIS.esriSystem;
using ESRI.ArcGIS.SystemUI;

namespace DynamicBiking
{
	/// <summary>
	/// Summary description for BikingTrackModeCmd.
	/// </summary>
	[Guid("5a26e262-9e4c-498f-b77c-a6fdeee0dd4b")]
	[ClassInterface(ClassInterfaceType.None)]
	[ProgId("DynamicBiking.BikingTrackModeCmd")]
	public sealed class BikingTrackModeCmd : BaseCommand
	{
		#region COM Registration Function(s)
		[ComRegisterFunction()]
		[ComVisible(false)]
		static void RegisterFunction(Type registerType)
		{
			// Required for ArcGIS Component Category Registrar support
			ArcGISCategoryRegistration(registerType);

			//
			// TODO: Add any COM registration code here
			//
		}

		[ComUnregisterFunction()]
		[ComVisible(false)]
		static void UnregisterFunction(Type registerType)
		{
			// Required for ArcGIS Component Category Registrar support
			ArcGISCategoryUnregistration(registerType);

			//
			// TODO: Add any COM unregistration code here
			//
		}

		#region ArcGIS Component Category Registrar generated code
		/// <summary>
		/// Required method for ArcGIS Component Category registration -
		/// Do not modify the contents of this method with the code editor.
		/// </summary>
		private static void ArcGISCategoryRegistration(Type registerType)
		{
			string regKey = string.Format("HKEY_CLASSES_ROOT\\CLSID\\{{{0}}}", registerType.GUID);
			ControlsCommands.Register(regKey);

		}
		/// <summary>
		/// Required method for ArcGIS Component Category unregistration -
		/// Do not modify the contents of this method with the code editor.
		/// </summary>
		private static void ArcGISCategoryUnregistration(Type registerType)
		{
			string regKey = string.Format("HKEY_CLASSES_ROOT\\CLSID\\{{{0}}}", registerType.GUID);
			ControlsCommands.Unregister(regKey);

		}

		#endregion
		#endregion

		private IHookHelper m_hookHelper = null;

		private DynamicBikingCmd m_dynamicBikingCmd = null;

		public BikingTrackModeCmd()
		{
			base.m_category = ".NET Samples";
			base.m_caption = "Dynamic biking track";
			base.m_message = "Dynamic biking track mode";
			base.m_toolTip = "Dynamic biking track mode";
			base.m_name = "DynamicBiking_BikingTrackModeCmd";

			try
			{
				string bitmapResourceName = GetType().Name + ".bmp";
				base.m_bitmap = new Bitmap(GetType(), bitmapResourceName);
			}
			catch (Exception ex)
			{
				System.Diagnostics.Trace.WriteLine(ex.Message, "Invalid Bitmap");
			}
		}

		#region Overriden Class Methods

		/// <summary>
		/// Occurs when this command is created
		/// </summary>
		/// <param name="hook">Instance of the application</param>
		public override void OnCreate(object hook)
		{
			if (hook == null)
				return;

			if (m_hookHelper == null)
				m_hookHelper = new HookHelperClass();

			m_hookHelper.Hook = hook;
		}

		/// <summary>
		/// Occurs when this command is clicked
		/// </summary>
		public override void OnClick()
		{
			if (m_dynamicBikingCmd != null)
			{
				m_dynamicBikingCmd.TrackMode = !m_dynamicBikingCmd.TrackMode;
			}
		}

		public override bool Enabled
		{
			get
			{
				m_dynamicBikingCmd = GetBikingCmd();
				if (m_dynamicBikingCmd != null)
					return m_dynamicBikingCmd.IsPlaying;

				return false;
			}
		}

		public override bool Checked
		{
			get
			{
				if (m_dynamicBikingCmd != null)
				{
					return m_dynamicBikingCmd.TrackMode;
				}
				return false;
			}
		}

		#endregion

		private DynamicBikingCmd GetBikingCmd()
		{
			if (m_hookHelper.Hook == null)
				return null;

			DynamicBikingCmd dynamicBikingCmd = null;
			if (m_hookHelper.Hook is IToolbarControl2)
			{
				IToolbarControl2 toolbarCtrl = (IToolbarControl2)m_hookHelper.Hook;
				ICommandPool2 commandPool = toolbarCtrl.CommandPool as ICommandPool2;
				int commantCount = commandPool.Count;
				ICommand command = null;
				for (int i = 0; i < commantCount; i++)
				{
					command = commandPool.get_Command(i);
					if (command is DynamicBikingCmd)
					{
						dynamicBikingCmd = (DynamicBikingCmd)command;
					}
				}
			}

			return dynamicBikingCmd;
		}
	}
}
